### PR TITLE
Switch to common/{model,expfmt}

### PR DESCRIPTION
**Pull request by fabxc**
https://github.com/prometheus/common

With this support for the legacy JSON formats gets dropped.

### CODE REVIEW

**Comment by fabxc** [comment]
Federation now using common/expfmt, too.
Prometheus now longer uses client_golang/{extraction,model,text}. Dependencies to those remain through the instrumentation library and will disappear as soon as client_golang/prometheus also switches to the common packages.

@juliusv

**Comment by brian-brazil** [comment]
> Dependencies to those remain through the instrumentation library and will disappear as soon as client_golang/prometheus also switches to the common packages.

Offtopic for this PR but I'm a bit worried about the instrumentation bit of the client having any dependencies outside of itself for anything that doesn't come with Go, as we need it as light as possible so there's less resistance to using it everywhere. A dependency within the same repo may be okay, but I don't think another repository (particularity what's looks at first glance to be a part of the server) is something we should be having.

**Comment by juliusv** [comment]
@brian-brazil So you are worried about the client library depending on https://github.com/prometheus/common? That's better IMO though than keeping all the shared non-clientlib functionality in client_golang.

**Comment by fabxc** [comment]
There's absolutely no technical difference regarding where a package comes from. Also the instrumentation library has dependencies to three different repos that don't even belong to the Prometheus organization.

The common repo clearly separates our model and exposition format implementation (which are internal) from the versioning of the instrumentation library, which would bite us sooner or later with the current state. It's under our control so this is nothing to worry about.
I want to address the dependencies on the truly external repositories in the long run.

The instrumentation library is simply to complex to have it all in a single package without becoming unmaintainable – I assure you that I've thought about it extensively. Eventually I want to move towards two dependencies: `common/{model,expfmt}`, which is a significant improvement from the current state.

It's worth noting that there haven't even been complaints about the current state.

File: retrieval/target_test.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	"errors"
-	"fmt"
+	// "fmt"
```
Comment (juliusv) [inline retrieval/target_test.go]: Remove commented-out line.

File: retrieval/target_test.go [hunk 2]
```diff
@@ -140,31 +140,31 @@ func TestTargetScrapeUpdatesState(t *testing.T) {
...
-
-	testTarget.scrape(slowAppender{})
-	if testTarget.status.Health() != HealthBad {
-		t.Errorf("Expected target state %v, actual: %v", HealthBad, testTarget.status.Health())
-	}
-	if testTarget.status.LastError() != errIngestChannelFull {
-		t.Errorf("Expected target error %q, actual: %q", errIngestChannelFull, testTarget.status.LastError())
-	}
-}
+// func TestTargetScrapeWithFullChannel(t *testing.T) {
```
Comment (juliusv) [inline retrieval/target_test.go]: Status of this flaky test? We still don't know why it's even flaky, right?

Comment (fabxc) [inline retrieval/target_test.go]: Ah yes, just commented this out. As the decoder removes the indirection over a channel this test does  no longer make sense as it is.

I will see whether we need any sort of replacement at all.

**Comment by brian-brazil** [comment]
> @brian-brazil So you are worried about the client library depending on https://github.com/prometheus/common? That's better IMO though than keeping all the shared non-clientlib functionality in client_golang.

Yes, I'm worried about anything that makes the client library not self contained.

> There's absolutely no technical difference regarding where a package comes from. Also the instrumentation library has dependencies to three different repos that don't even belong to the Prometheus organization. I want to address the dependencies on the truly external repositories in the long run.

There's no technical difference, but it prevents wider usage of the library. Removing those dependencies is good.

> Eventually I want to move towards two dependencies: common/{model,expfmt}, which is a significant improvement from the current state.

expfmt shouldn't be a dependency of the instrumentation library, as exposition should be separate.

> It's worth noting that there haven't even been complaints about the current state.

I've seen complaints for the Go library, and similar complaints about the original Java library. Protobuf is not a light dependency for example. My vision is that our client libraries are the default choice for instrumentation in all languages, this is one barrier to that.

**Comment by fabxc** [comment]
I could see us dropping the dependency on any model package. This is part of a later general refactoring of the instrumentation library. So are steps to possibly dropping a protobuf dependency.

To be clear – minimizing dependencies is a clear goal but zero is just neither realistic, nor necessary. Instrumentation is simply not a small thing to do, it comes with a cost.
Large projects integrated our library in the current state without complaint (e.g. Kubernetes, etcd). This will only improve with a standardized vendoring approach that's being introduced with Go 1.5.

This a first step to improve the situation significantly in the long run. Your concerns are not related to it. Removing the extraction/text/model packages from the versioning of the client_golang library is the right thing to do.

**Comment by juliusv** [comment]
@brian-brazil I agree that exposition should be pluggable, but default mechanisms should be included, no? Otherwise the user always needs to use two packages to get metrics exported.

Btw., we already have a several smaller external dependencies besides protobufs: perks,  HTTP negotiation, and delimited protobufs for example. Over time, we might want to integrate them into our code directly.

As for protobufs, they are not only used as an exposition format, but as the canonical internal representation format of samples within the client library. So not depending on protobuf would require a bigger reworking of the internals, and would lead to less efficiency as the protobuf exposition case would then require one more decoding/encoding step.

**Comment by juliusv** [comment]
:+1: 

Things to note:
- I haven't fully reviewed the vendored `common` package itself yet, but its usage looks good.
- We may want to fix the import grouping if feasible...

**Comment by brian-brazil** [comment]
:+1: on this change itself.

> To be clear – minimizing dependencies is a clear goal but zero is just neither realistic, nor necessary. 

I think it's realistic and necessary, I've managed it with both the Python client and Java instrumentation libraries due to user hesitance to even consider Prometheus.

> @brian-brazil I agree that exposition should be pluggable, but default mechanisms should be included, no? 

Exposition is separate from instrumentation. If a client library makes that distinction dependency wise (java does, python doesn't) then instrumentation shouldn't depend on exposition.

> Otherwise the user always needs to use two packages to get metrics exported.

That's true for an application, but not a library. Usually where you're doing the exposition bits is not the same place you're doing instrumentation.

> As for protobufs, they are not only used as an exposition format, but as the canonical internal representation format of samples within the client library. So not depending on protobuf would require a bigger reworking of the internals, and would lead to less efficiency as the protobuf exposition case would then require one more decoding/encoding step.

I had presumed that that's what main part of the planned rework would do, presuming we can't inline it.

> Btw., we already have a several smaller external dependencies besides protobufs: perks, HTTP negotiation, and delimited protobufs for example. Over time, we might want to integrate them into our code directly.

That's exposition, which it's okay to have dependencies for. Having the default avoid something heavy (and contentious in some parts) like protobuf may be an idea, depending on what the state of Go dependency management is.

**Comment by juliusv** [comment]
@brian-brazil Good point on exposition being in a different place of the code. Yeah, that would make it easier to make the user use a different package then.

As for protobufs, AFAIR @beorn7's plan was to keep them as the canonical internal representation even after a rework, but you might want to check with him directly after he comes back next Wednesday.

**Comment by fabxc** [comment]
We can see how far we get in reducing dependencies. I think 'necessity' greatly depends on how much a language's eco system is broken in this regard.

**Comment by brian-brazil** [comment]
It's not just the operational aspects of dependency management, there also a desire to avoid dependencies you don't need. For example if I'm using the prometheus client to only expose to graphite or influxdb, why should I need to pull in protobuf?

**Comment by fabxc** [comment]
I agree, I'm sure we can work something out. The current situation certainly needs improvement.
I'm just strongly against stuffing things that clearly belong into separate packages into a single one for the sake of having 'no dependencies' (which is of course a lie because I gave up clear separation of concerns and stuff all the code into a single folder with prefixed file names).
If we can design the whole thing in such a way that this is of no concern, even better.

**Comment by juliusv** [comment]
:+1:

**Comment by beorn7** [comment]
FTR: In a way, 'metrics.proto` is the core of Prometheus (ask Matt P. about historical details and the original vision – which is in important parts different from the often mentioned Borgmon). The fact that the server breaks metrics as represented by the protobufs into individual one-dimensional time series is an implementation detail. It's coincidence (or convenience) that the line-based text format uses the same scheme. A future, better Prometheus server will be able to natively handle more complex metrics than "just gauges" and keep track (and make use) of all the meta data. Obviously, nothing forces us to stick to protobuf as the format for the canonical representation of Prometheus metrics. We could change it, but for now, it is protobuf. The line-based text format is _not_ suitable as the canonical representation.

From a fundamental design point-of-view, it is desirable that both client and server share the canonical representation of Prometheus metrics. The fact that we have client libraries now that are internally ignorant of the canonical representation deeply worries me. I understand the reasons why they were done that way, and they are still linked to the canonical representation via the parser of the line-based text format (which recreates the canonical representation albeit in an arcane way that relies deeply on conventions).

To keep us in a position where we can make use of all the aspects of the canonical representation (as originally envisioned by our founders), we need to keep it in the loop, i.e. the server _must_ convert everything that is scraped into the canonical format first (and only then apply internal conversions into individual time series and such) and clients _should_ be able to internally use the canonical representation.

The line-based text format is a compromise to allow simple or even ad-hoc implementation of clients. But it is a maintenance burden (as we have to keep track of conversion from the text format into the canonical representation (once, in Go) and vice versa (many times, for each language that uses the text format).

**Comment by fabxc** [comment]
:+1: 
Also, no streaming with text format, should that ever become desirable.

**Comment by brian-brazil** [comment]
From a marketing and usage standpoint, we need to remove the dependency on protobuf from the instrumentation client - it puts people off, gets us into religious wars that we don't need to (some people care a bit too much about interchange formats) and ultimately prevents the adoption of our instrumentation libraries into other systems. If we can get rid of the hard requirement for protobuf for exposition that'd also be good.

Having protobuf limits our ability to grow the community and improve monitoring for everyone. We'd have the exact same problem if we were using something like Thrift, XML or Cap'n'proto.

It doesn't matter what we use inside the server, what's inside the clients does matter though and has already caused us problems with adoption.

**Comment by fabxc** [comment]
I still haven't heard a single person complaining about dependencies of the Go client.
Can we stop projecting the brokenness of one language on another.

> Having protobuf limits our ability to grow the community and improve monitoring for everyone

That's just a massive over-dramatization. 99% of people don't even care in either direction – to this day I haven't had a single conversation where protobuf was considered a flaw in our system in any way.

I still agree that separation should be a goal for our client libraries so the 1% who care can take a look and 5% of those that dislike protobuf can opt to not use it.
It's a good wire format. I strongly disagree with actively un-supporting it.

**Comment by brian-brazil** [comment]
> I still haven't heard a single person complaining about dependencies of the Go client.

@peterbourgon complained about the complexity and size of it on IRC a few months back.

> Can we stop projecting the brokenness of one language on another.

It's not just about languages, it's about users managing their dependencies and avoiding protobuf being an excuse or source of discord that prevents it being used. Taking it an extreme, what would you think of an instrumentation library that pulled in sqllite?

> That's just a massive over-dramatization. 99% of people don't even care in either direction – to this day I haven't had a single conversation where protobuf was considered a flaw in our system in any way.

I disagree, the few people who care are enough to prevent adoption. I have had lengthy discussions with those who saw protobuf as a problem.

It's not that it's a flaw, it's where it fits in the bigger picture of a potential user's ecosystem.

> I still agree that separation should be a goal for our client libraries so the 1% who care can take a look and the 5% of them that dislike protobuf can opt to not use it.

I think we need to have the default instrumentation library have no protobuf dependencies, with adding it in being easy (an extra import line).

We can't have someone doing some simple instrumentation pulling in protobuf by default, as that risks that PR descending into the advantages/disadvantages of protobuf rather than of prometheus instrumentation.

> It's a good wire format. I strongly disagree with actively un-supporting it.

I agree.

**Comment by fabxc** [comment]
> Having protobuf limits our ability to grow the community and improve monitoring for everyone. 

This threw me off as it sounded like we should get rid of protobuf altogether.

Now it sounds again like the end result of the last iteration of this discussion – after which I considered things resolved.

So let's make a final statement on this:
- We have different opinions regarding usefulness of protobufs...
- ...they don't matter because we agree that client libraries should provide clear separation between instrumentation and exposition making each format effectively opt-in
- Clients for different languages should provide exposition for protobuf and text format alike

Is that something we can all agree on as the end-goal?

**Comment by brian-brazil** [comment]
Add that instrumentation shouldn't depend on protobuf and I'll agree.

**Comment by beorn7** [comment]
The Go client fundamentally depends on protobuf because doing so is the cleanest approach.

Once the dust has settled around the ways of managing dependencies in Go, I could imagine there will be a way to make the protobuf usage completely internal within the Go client, i.e. dependency is vendored and will not interfere with other protobuf usage in the code that links in the Prometheus Go client. [1]

I'd be very reluctant to write badly designed software to drive adoption. In the given situation, I would not write a Go client that does not use protobuf.



[1] We tried vendoring within the library before but failed. At that time, we decided dependency management is a mess and we'll wait for a better solution in the future. I have several ideas how it could work but whatever we do should play well with approaches the Go community settles on, so it would be a bit early at this time.

**Comment by brian-brazil** [comment]
> dependency is vendored and will not interfere with other protobuf usage in the code that links in the Prometheus Go client. 

To be clear the issue isn't vendoring (though we also need to solve that). As long as I pull down the go prometheus instrumentation client and it creates a directory somewhere called 'protobuf' or 'proto' this problem exists.

Whatever about protobuf's internal use within prometheus, it's a liability for the instrumentation client in terms of adoption.

> I'd be very reluctant to write badly designed software to drive adoption. In the given situation, I would not write a Go client that does not use protobuf.

I appreciate that this is your opinion and I'm also a fan of protobuf -  but many potential users out there have for whatever the exact opposite opinion of protobuf. Even bringing the question to the table of what interchange format to use in the instrumentation client means we have already lost, as that has the potential to start a religious war.

In terms of a pure instrumentation library, I don't see a benefit of protobuf over a struct. It's only when we get to exposition that it comes in, and I hope to have many non-protobuf and non-prometheus expositions in the future. If we want to get applications and libraries directly instrumented, we need to make the choice of the Prometheus instrumentation library as easy as possible and remove potential objections to it.

This isn't about saying what we have is good or bad, but about what practical steps we need to take to grow the ecosystem.

**Comment by peterbourgon** [comment]
As I was @-mentioned, I'll quickly chime in — I am indeed put off by the number of (transient) dependencies pulled in by client_golang, but not terribly so, and vendoring those dependencies would be a fine solution, in the long term.

> As long as I pull down the go prometheus instrumentation client and it creates a directory somewhere called 'protobuf' or 'proto' this problem exists.

Given the implementation detail of protobuf never leaks beyond the instrumentation package boundary, that's quite an unreasonable position. Not sure if serious.

**Comment by beorn7** [comment]
Speaking about Go specifically: It is not a problem to vendor the protobuf dependency under a different import path. That's what we did in our past approach. In that way, our "internal" protobuf code didn't interfere with protobuf code imported by the client code that was using `client_golang`. The problem was compatibility with Prometheus-related protobuf types that were not renamed in the same way. Since protobuf types appear in function signatures of `client_golang`, collisions can easily happen. With some care, we could likely come up with a less problematic way of vendoring. In a way, all references to protobuf types would appear as "some struct" to the user, with no relation to protobuf types they have imported themself.

**Comment by brian-brazil** [comment]
> Given the implementation detail of protobuf never leaks beyond the instrumentation package boundary, 

It does leak.

> that's quite an unreasonable position. Not sure if serious.

I think it's a perfectly reasonable situation if you're in a situation when you're trying to keep your dependencies under control, either to manage bloat and/or reduce maintenance down the line. 

There are people out there with this position, I have interacted with them on many occasions - it's pretty frustrating to do so and I get the sense that they're used to having to reject many spurious new dependencies so that dependencies that are justified are also hard to get by them.

There are also those who react negatively to protobuf, and a single strong objector within a project could be sufficient to nix a prometheus integration.

The question we have to ask ourselves is does having protobuf as a hard dependency in instrumentation make the instrumentation so much better that we're willing to lose these integrations? As it stands protobuf offers no advantage to the average person doing direct instrumentation (as it's hidden from them), but loses us potential users. In Go we're already competing with several other libraries for instrumentation, let's not make that harder than it has to be.

> In a way, all references to protobuf types would appear as "some struct" to the user, with no relation to protobuf types they have imported themself.

That's one approach that I think would work. Having a quick look at the instrumentation code we're actually not that tightly bound to protobuf so having the dependency only in the protobuf exposition code should be doable, SetMetricFamilyInjectionHook is the only real wrinkle (and we control both places it's used).

**Pull request merged**